test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyProjectReferenceCompat.cs

```diff
@@ -0,0 +1,127 @@
+        //  Fullframework NuGet versioning on Jenkins infrastructure issue
+        //        https://github.com/dotnet/sdk/issues/1041
+        //[InlineData("netcoreapp2.0", "FullMatrix", "netstandard1.0 netstandard1.1 netstandard1.2 netstandard1.3 netstandard1.4 netstandard1.5 netstandard1.6 netstandard2.0 netcoreapp1.0 netcoreapp1.1 netcoreapp2.0", true)]
+
+        public void It_checks_for_valid_project_reference_compat(string referencerTarget, string testIDPostFix, string rawDependencyTargets, bool buildSucceeds)
+        {
+            string identifier = "_TestID_" + referencerTarget + "_" + testIDPostFix;
+
+            TestProject referencerProject = GetTestProject("Referencer", referencerTarget, true);
+            List<string> dependencyTargets = rawDependencyTargets.Split(',', ';', ' ').ToList();
+            int dependencyTargetNamingIndex = 1;
+            foreach (string dependencyTarget in dependencyTargets)
+            {
+                TestProject dependencyProject = GetTestProject("Dependency" + dependencyTargetNamingIndex++, dependencyTarget, true);
+                referencerProject.ReferencedProjects.Add(dependencyProject);
+            }
+
+            //  Set the referencer project as an Exe unless it targets .NET Standard
+            if (!referencerProject.ShortTargetFrameworkIdentifiers.Contains("netstandard"))
+            {
+                referencerProject.IsExe = true;
+            }
+
+            //  Skip running .NET Framework tests if not running on Windows
+            //        https://github.com/dotnet/sdk/issues/335
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                if (!AllProjectsBuildOnNonWindows(referencerProject))
+                {
+                    return;
+                }
+            }
+
+            var testAsset = _testAssetsManager.CreateTestProject(referencerProject, nameof(It_checks_for_valid_project_reference_compat), identifier);
+            var restoreCommand = testAsset.GetRestoreCommand(relativePath: referencerProject.Name).Execute().Should().Pass();
+            var appProjectDirectory = Path.Combine(testAsset.TestRoot, referencerProject.Name);
+
+            var buildCommand = new BuildCommand(Stage0MSBuild, appProjectDirectory);
+            if (!buildSucceeds)
+            {
+                buildCommand = buildCommand.CaptureStdOut();
+            }
+
+            var result = buildCommand.Execute();
+
+            if (buildSucceeds)
+            {
+                result.Should().Pass();
+            }
+            else
+            {
+                result.Should().Fail()
+                    .And.HaveStdOutContaining("It cannot be referenced by a project that targets");
+            }
+        }
+
+        TestProject GetTestProject(string name, string target, bool isSdkProject)
+        {
+            TestProject ret = new TestProject()
```

Comment:
nit 1: Please call this something more descriptive than `ret`. `testProject` would work fine. 
nit 2: Use `var` when type is obvious

``` C#
var testProject = new TestProject()
```